relay-general/src/processor/attrs.rs

```diff
@@ -212,6 +213,61 @@ pub struct FieldAttrs {
     pub retain: bool,
 }
 
+/// A set of characters allowed or denied for a (string) field.
+///
+/// Note that this field is generated in the derive, it can't be constructed easily in tests.
+#[derive(Clone, Copy)]
+pub struct CharacterSet {
+    /// Generated in derive for performance. Can be left out when set is created manually.
+    pub is_match: Option<fn(char) -> bool>,
+    /// A set of ranges that are allowed/denied within the character set
+    pub ranges: &'static [RangeInclusive<char>],
+    /// Whether the character set is inverted
+    pub is_negative: bool,
```

Comment:
Do you need the `is_negative` flag just for matching or also for the JSON schema. If you only need it for matching, you can invert the condition in the function instead.